src/rendezvous/MSC3906Rendezvous.ts

```diff
@@ -0,0 +1,269 @@
+    Finish = 'm.login.finish',
+    Progress = 'm.login.progress',
+}
+
+enum Outcome {
+    Success = 'success',
+    Failure = 'failure',
+    Verified = 'verified',
+    Declined = 'declined',
+    Unsupported = 'unsupported',
+}
+
+interface RendezvousPayload {
+    type: PayloadType;
+    intent?: RendezvousIntent;
+    outcome?: Outcome;
+    device_id?: string;
+    device_key?: string;
+    verifying_device_id?: string;
+    verifying_device_key?: string;
+    master_key?: string;
+    protocols?: string[];
+    protocol?: string;
+    login_token?: string;
+    homeserver?: string;
+}
+
+const LOGIN_TOKEN_PROTOCOL = new UnstableValue("login_token", "org.matrix.msc3906.login_token");
+
+/**
+ * Implements MSC3906 to allow a user to sign in on a new device using QR code.
+ * This implementation only supports generating a QR code on a device that is already signed in.
+ * Note that this is UNSTABLE and may have breaking changes without notice.
+ */
+export class MSC3906Rendezvous {
+    private newDeviceId?: string;
+    private newDeviceKey?: string;
+    private ourIntent: RendezvousIntent = RendezvousIntent.RECIPROCATE_LOGIN_ON_EXISTING_DEVICE;
+    private _code?: string;
+
+    /**
+     * @param channel The secure channel used for communication
+     * @param client The Matrix client in used on the device already logged in
+     * @param onFailure Callback for when the rendezvous fails
+     */
+    public constructor(
+        private channel: RendezvousChannel,
+        private client: MatrixClient,
+        public onFailure?: RendezvousFailureListener,
+    ) {}
+
+    /**
+     * Returns the code representing the rendezvous suitable for rendering in a QR code or undefined if not generated yet.
+     */
+    public get code(): string | undefined {
+        return this._code;
+    }
+
+    /**
+     * Generate the code including doing partial set up of the channel where required.
+     */
+    public async generateCode(): Promise<void> {
+        if (this._code) {
+            return;
+        }
+
+        this._code = JSON.stringify(await this.channel.generateCode(this.ourIntent));
+    }
+
+    public async startAfterShowingCode(): Promise<string | undefined> {
+        const checksum = await this.channel.connect();
+
+        logger.info(`Connected to secure channel with checksum: ${checksum} our intent is ${this.ourIntent}`);
+
+        const features = await buildFeatureSupportMap(await this.client.getVersions());
+        // determine available protocols
+        if (features.get(Feature.LoginTokenRequest) === ServerSupport.Unsupported) {
+            logger.info("Server doesn't support MSC3882");
+            await this.send({ type: PayloadType.Finish, outcome: Outcome.Unsupported });
+            await this.cancel(RendezvousFailureReason.HomeserverLacksSupport);
+            return undefined;
+        }
+
+        await this.send({ type: PayloadType.Progress, protocols: [LOGIN_TOKEN_PROTOCOL.name] });
```

Comment:
It is strange that this call mixes stable and unstable-prefixed identifiers, both which are introduced in the same changeset, this seems wrong

Reply:
Well, one could argue that everything needs unstable-prefixes as everything is new. I think that is overkill.

Reply:
That's not necessarily the case. If the API you are sending on has an unstable prefix part of the same MSC then all identifiers within can be stable. Otherwise all new identifiers must be unstable.